### DISCUSSION TITLE
New structure

### DISCUSSION THREAD

**Discussion by VForiel** [Gunivers/Gipsy#31] [General]
Here is a proposition of structure for Gipsy 2.0
I tried to make it as simple as possible by taking every needs and future propects we discussed until now. Feel free to give your feedback.

# ACTIVE PR: #45 

# Bot Sturcture
```
📄 start.py
📄 setup.py
📄 config.py
📄 requirements.txt
📄 README.md
📄 LICENSE.txt
📁 data
└─ 📄data.db
📁 core
├─ 📄 sconfig.py
├─ 📄 i18n.py
├─ 📄 log.py
└─ 📄  ... other core systems
📁 logs
└─ 📄  ... log files
📁 plugins
└─ 📁 <plugin_name>
   └─ ... see below
```

# Plugin structure
```
📁 <plugin_name>
├─ 📄 <plugin_name>.py // cog initialization, command definition etc.
├─ 📄 config.py // this file will be copied in the main config folder during the installation process
├─ 📄 database.sql // will be removed once the database will be automatically managed using object attributes
├─ 📁 src
│  │  # plugin's additional fonctionnalities (python)
│  ├─ 📄 feature_1.py
│  └─ 📄 ...
├─ 📁 docs
│  │  # plugin documentation files (Markdown)
│  ├─ 📄 user_documentation.md
│  └─ 📄 ...
├─ 📁 langs
│  │  # plugin language files (YML)
│  ├─ 📄 en.yml
│  ├─ 📄 fr.yml
│  └─ 📄 ...
├─ 📁 api
   │  # plugin API (python)
   └─ 📄 /!\ This part is not defined yet. It's a future prospect.
└─ 📁 web
   │  # plugin web UI files (JS)
   └─ 📄 /!\ This part is not defined yet. It's a future prospect.
```

**Comment by ascpial** [comment]
As I can see in the data directory, you want to do a fragmented database?
I think this is not a good idea for performance reasons, and also some systems might need cross-server data storage such as the wormhole system.

**Reply by VForiel** [reply]
Good point. The idea was to be able to move the data of a given server from a bot to another (if a server decide to self-host the bot) but there is probably better solution to make it possible by keeping the database unified

**Reply by ascpial** [reply]
I think this type of feature can be done without the need of designing the bot for it.
The server owner will always need to get the data so I think we can just make a copy command that create a copy of the database concerning the server that can be imported to a self-hosted instance.

**Reply by Aeris1One** [reply]
getting the data is an obligation of the GDPR (RGPD in the Molière's langue) : as the bot gains data about servers and people, those servers and people have to be able to retrieve this data, and optionally delete it
